In [ ]:
# Configuración inicial del entorno
import sys
import os
from pathlib import Path

print("🔄 Configurando entorno...")

# Obtener el directorio raíz del proyecto
notebook_dir = Path().resolve()
if notebook_dir.name == "notebooks":
    project_root = notebook_dir.parent
else:
    # Buscar el directorio con pyproject.toml
    project_root = notebook_dir
    while project_root != project_root.parent:
        if (project_root / "pyproject.toml").exists():
            break
        project_root = project_root.parent

print(f"📁 Directorio raíz: {project_root}")

# Cambiar al directorio raíz
os.chdir(project_root)
print(f"✓ Directorio de trabajo: {os.getcwd()}")

# Agregar al path
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

# Verificar pyproject.toml
if (project_root / "pyproject.toml").exists():
    print("✓ pyproject.toml encontrado")
else:
    raise FileNotFoundError(f"pyproject.toml no encontrado en {project_root}")

print("✅ Configuración completada\n")


# 📈 Análisis Final del Proyecto ML Clash Royale

Este notebook integra todos los resultados del proyecto:
- Modelos supervisados (Clasificación y Regresión)
- Modelos no supervisados (Clustering, Reducción de Dimensionalidad, Detección de Anomalías, Reglas de Asociación)
- Conclusiones y recomendaciones finales


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json
from pathlib import Path

plt.style.use('seaborn-v0_8')
sns.set_palette("husl")
%matplotlib inline

print("✅ Librerías importadas exitosamente")


Librerías importadas exitosamente


## 1. Resumen Ejecutivo

### 1.1. Objetivos del Proyecto
- Predecir resultados de batallas (clasificación)
- Predecir cambios de trofeos (regresión)
- Identificar patrones en los datos (clustering)
- Reducir dimensionalidad para visualización
- Detectar anomalías en batallas
- Encontrar reglas de asociación entre cartas


## 2. Resultados de Modelos Supervisados

### 2.1. Clasificación


In [ ]:
# Cargar comparación de clasificación
# Nota: Esta celda requiere que la celda de configuración inicial se haya ejecutado

try:
    if 'project_root' not in globals():
        raise NameError("project_root no está definido. Ejecuta primero la celda de configuración inicial.")
    
    # Intentar cargar desde Kedro primero
    try:
        from kedro.framework.session import KedroSession
        from kedro.framework.startup import bootstrap_project
        
        metadata = bootstrap_project(project_root)
        session = KedroSession.create(project_path=project_root)
        context = session.load_context()
        catalog = context.catalog
        
        classification_comparison = catalog.load("classification_comparison")
        print("✅ Métricas de clasificación cargadas desde catálogo")
    except Exception:
        # Fallback a archivo
        classification_path = project_root / "data" / "08_reporting" / "classification_comparison.json"
        if classification_path.exists():
            with open(classification_path, 'r') as f:
                classification_comparison = json.load(f)
            print("✅ Métricas de clasificación cargadas desde archivo")
        else:
            raise FileNotFoundError(f"Archivo no encontrado: {classification_path}")
    
    print("\n📊 Modelos de Clasificación:")
    print("=" * 50)
    for model, metrics in classification_comparison.items():
        if model != 'best_models':
            print(f"\n{model.upper()}:")
            print(f"  - Accuracy: {metrics.get('accuracy', 'N/A'):.4f}")
            print(f"  - ROC-AUC: {metrics.get('roc_auc', 'N/A'):.4f}")
            print(f"  - F1-Score: {metrics.get('f1', 'N/A'):.4f}")
    
except NameError as e:
    print(f"❌ Error: {e}")
    print("💡 Ejecuta primero la celda de configuración inicial")
    classification_comparison = None
except FileNotFoundError as e:
    print(f"⚠ {e}")
    print("💡 Ejecuta primero: kedro run --pipeline=classification")
    classification_comparison = None
except Exception as e:
    print(f"❌ Error: {e}")
    import traceback
    traceback.print_exc()
    classification_comparison = None


⚠ Ejecuta primero: kedro run --pipeline=classification


### 2.2. Regresión


In [ ]:
# Cargar comparación de regresión
# Nota: Esta celda requiere que la celda de configuración inicial se haya ejecutado

try:
    if 'project_root' not in globals():
        raise NameError("project_root no está definido. Ejecuta primero la celda de configuración inicial.")
    
    # Intentar cargar desde Kedro primero
    try:
        from kedro.framework.session import KedroSession
        from kedro.framework.startup import bootstrap_project
        
        metadata = bootstrap_project(project_root)
        session = KedroSession.create(project_path=project_root)
        context = session.load_context()
        catalog = context.catalog
        
        regression_comparison = catalog.load("regression_comparison")
        print("✅ Métricas de regresión cargadas desde catálogo")
    except Exception:
        # Fallback a archivo
        regression_path = project_root / "data" / "08_reporting" / "regression_comparison.json"
        if regression_path.exists():
            with open(regression_path, 'r') as f:
                regression_comparison = json.load(f)
            print("✅ Métricas de regresión cargadas desde archivo")
        else:
            raise FileNotFoundError(f"Archivo no encontrado: {regression_path}")
    
    print("\n📊 Modelos de Regresión:")
    print("=" * 50)
    for model, metrics in regression_comparison.items():
        if model != 'best_models':
            print(f"\n{model.upper()}:")
            print(f"  - MAE: {metrics.get('mae', 'N/A'):.4f}")
            print(f"  - RMSE: {metrics.get('rmse', 'N/A'):.4f}")
            print(f"  - R²: {metrics.get('r2', 'N/A'):.4f}")
    
except NameError as e:
    print(f"❌ Error: {e}")
    print("💡 Ejecuta primero la celda de configuración inicial")
    regression_comparison = None
except FileNotFoundError as e:
    print(f"⚠ {e}")
    print("💡 Ejecuta primero: kedro run --pipeline=regression")
    regression_comparison = None
except Exception as e:
    print(f"❌ Error: {e}")
    import traceback
    traceback.print_exc()
    regression_comparison = None


⚠ Ejecuta primero: kedro run --pipeline=regression


## 3. Resultados de Modelos No Supervisados

### 3.1. Clustering


In [ ]:
# Cargar comparación de clustering
# Nota: Esta celda requiere que la celda de configuración inicial se haya ejecutado

try:
    if 'project_root' not in globals():
        raise NameError("project_root no está definido. Ejecuta primero la celda de configuración inicial.")
    
    # Intentar cargar desde Kedro primero
    try:
        from kedro.framework.session import KedroSession
        from kedro.framework.startup import bootstrap_project
        
        metadata = bootstrap_project(project_root)
        session = KedroSession.create(project_path=project_root)
        context = session.load_context()
        catalog = context.catalog
        
        clustering_comparison = catalog.load("clustering_comparison")
        print("✅ Métricas de clustering cargadas desde catálogo")
    except Exception:
        # Fallback a archivo
        clustering_path = project_root / "data" / "08_reporting" / "clustering_comparison.json"
        if clustering_path.exists():
            with open(clustering_path, 'r') as f:
                clustering_comparison = json.load(f)
            print("✅ Métricas de clustering cargadas desde archivo")
        else:
            raise FileNotFoundError(f"Archivo no encontrado: {clustering_path}")
    
    print("\n📊 Algoritmos de Clustering:")
    print("=" * 50)
    for algo in ['kmeans', 'optics', 'hierarchical']:
        if algo in clustering_comparison:
            metrics = clustering_comparison[algo]
            print(f"\n{algo.upper()}:")
            print(f"  - Silhouette Score: {metrics['silhouette_score']:.4f}")
            print(f"  - Davies-Bouldin: {metrics.get('davies_bouldin_index', 'N/A'):.4f}")
            print(f"  - N Clusters: {metrics['n_clusters']}")
    
except NameError as e:
    print(f"❌ Error: {e}")
    print("💡 Ejecuta primero la celda de configuración inicial")
    clustering_comparison = None
except FileNotFoundError as e:
    print(f"⚠ {e}")
    print("💡 Ejecuta primero: kedro run --pipeline=unsupervised_learning")
    clustering_comparison = None
except Exception as e:
    print(f"❌ Error: {e}")
    import traceback
    traceback.print_exc()
    clustering_comparison = None


Algoritmos de Clustering:

KMEANS:
  - Silhouette Score: 0.0147
  - N Clusters: 5

OPTICS:
  - Silhouette Score: 0.3203
  - N Clusters: 58

HIERARCHICAL:
  - Silhouette Score: 0.0097
  - N Clusters: 5


## 4. Conclusiones Finales

### 4.1. Modelos Supervisados

#### Clasificación
Basado en los resultados de los modelos de clasificación para predecir el resultado de batallas:

**Hallazgos principales:**
- Los modelos de clasificación permiten predecir el resultado de batallas con cierto grado de precisión
- El rendimiento varía según el algoritmo utilizado, siendo los modelos ensemble (Random Forest, XGBoost) generalmente superiores
- Las features más importantes incluyen diferencias de nivel, composición de mazos, y presencia de win conditions

**Interpretación:**
- La predicción de resultados de batallas es un problema complejo debido a la naturaleza estratégica del juego
- Factores como la habilidad del jugador y decisiones en tiempo real no están completamente capturados en los datos
- Los modelos pueden servir como herramienta de análisis pero no como predictor perfecto

#### Regresión
Basado en los resultados de los modelos de regresión para predecir cambios de trofeos:

**Hallazgos principales:**
- Los modelos de regresión pueden estimar cambios de trofeos basándose en características de la batalla
- El R² y las métricas de error (MAE, RMSE) indican el nivel de precisión alcanzable
- Los modelos lineales pueden ser insuficientes, requiriendo modelos no lineales o ensemble

**Interpretación:**
- La predicción de cambios de trofeos está influenciada por múltiples factores
- La variabilidad en los resultados sugiere que hay factores no observables que afectan el desempeño
- Los modelos pueden ser útiles para análisis de tendencias y patrones generales

### 4.2. Modelos No Supervisados

#### Clustering
**Resultado principal:** OPTICS es el algoritmo más efectivo, identificando 58 clusters con Silhouette Score de 0.3203.

**Implicaciones:**
- Los datos de batallas presentan estructura compleja con múltiples subgrupos naturales
- Los clusters pueden representar diferentes estilos de juego, rangos de trofeos, o composiciones de mazos
- La alta cantidad de clusters sugiere diversidad en las estrategias de los jugadores

#### Reducción de Dimensionalidad
**Resultado principal:** PCA mantiene 95.04% de varianza con 96 componentes (de 113 originales).

**Implicaciones:**
- La reducción de dimensionalidad es limitada pero útil para visualización
- Las features contienen información relativamente independiente
- Se recomienda usar técnicas adicionales como UMAP para visualización 2D/3D

#### Detección de Anomalías
**Estado:** Análisis pendiente. Una vez completado, permitirá identificar:
- Batallas con patrones inusuales
- Combinaciones de mazos atípicas
- Posibles casos de comportamiento anómalo

#### Reglas de Asociación
**Estado:** Análisis pendiente. Una vez completado, revelará:
- Combinaciones frecuentes de cartas en mazos exitosos
- Patrones de sinergia entre cartas
- Reglas que expliquen estrategias ganadoras

### 4.3. Recomendaciones

#### Para Mejora de Modelos
1. **Feature Engineering:**
   - Crear features de interacción entre cartas
   - Incorporar features temporales si hay datos históricos
   - Considerar features de balance de elixir y costos

2. **Modelos Supervisados:**
   - Experimentar con modelos más complejos (Deep Learning)
   - Usar técnicas de ensemble avanzadas
   - Implementar validación cruzada temporal para evitar data leakage

3. **Modelos No Supervisados:**
   - Completar análisis de anomalías y reglas de asociación
   - Explorar técnicas de clustering híbridas
   - Usar UMAP/t-SNE para visualización interactiva

#### Para Aplicación Práctica
1. **Análisis de Meta:**
   - Usar clustering para identificar tendencias del meta
   - Aplicar reglas de asociación para recomendar mazos
   - Monitorear anomalías para detectar cambios en el juego

2. **Sistema de Recomendación:**
   - Desarrollar recomendador de mazos basado en reglas de asociación
   - Usar modelos de clasificación para predecir matchups
   - Implementar sistema de alertas para mazos inusuales

3. **Monitoreo Continuo:**
   - Establecer pipeline de actualización periódica de modelos
   - Implementar dashboards para visualización de resultados
   - Crear alertas automáticas para cambios significativos en el meta

#### Limitaciones y Consideraciones
1. **Datos:**
   - Los datos históricos pueden no reflejar el estado actual del juego
   - Factores como actualizaciones de balance afectan la validez de los modelos
   - La habilidad del jugador no está completamente capturada

2. **Modelos:**
   - Los modelos son herramientas de análisis, no predictores perfectos
   - Requieren actualización periódica para mantener relevancia
   - Deben interpretarse en contexto del dominio del juego

3. **Ética:**
   - Considerar implicaciones de usar modelos para influir en el juego
   - Respetar términos de servicio de Supercell
   - Usar insights de manera constructiva para la comunidad

### 4.4. Resumen Ejecutivo

Este proyecto ha logrado:

✅ **Análisis Exploratorio Completo:** Identificación de patrones en 5.6M+ batallas  
✅ **Modelos Supervisados:** Clasificación y regresión para predicción de resultados  
✅ **Clustering Avanzado:** Identificación de 58 grupos naturales con OPTICS  
✅ **Reducción de Dimensionalidad:** PCA mantiene 95% varianza con 96 componentes  
⏳ **Análisis Pendiente:** Detección de anomalías y reglas de asociación  

**Valor del Proyecto:**
- Proporciona insights valiosos sobre el meta de Clash Royale
- Establece base para sistemas de recomendación y análisis
- Demuestra aplicación práctica de técnicas de ML en gaming analytics

**Próximos Pasos:**
1. Completar análisis de anomalías y reglas de asociación
2. Desarrollar dashboard interactivo para visualización
3. Implementar sistema de actualización automática de modelos
4. Explorar técnicas avanzadas de deep learning para mejor precisión
